File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/MIITMethodCountChecker.java
package com.loper7.miit_rule_checker;

import android.os.Handler;
import android.os.Looper;

import java.lang.reflect.Member;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

import top.canyie.pine.Pine;
import top.canyie.pine.callback.MethodHook;

/**
 * @Author loper7
 * @Date 2023/12/6 10:51
 * @Description
 **/
public class MIITMethodCountChecker {
    private static HashMap<String, HashMap<String, Integer>> map = new HashMap<>();
    private static Handler handler = new Handler(Looper.getMainLooper());


    /**
     * 开始计数
     *
     * @param members  需要计数的方法
     * @param deadline 截止时间
     */
    public static void startCount(Long deadline, ArrayList<Member> members) {
        map.clear();
        for (Member member : members) {
            if (member == null) continue;
            try {
                Pine.hook(member, new MethodHook() {
                    @Override
                    public void beforeCall(Pine.CallFrame callFrame) throws Throwable {
                        super.beforeCall(callFrame);
                        String methodName = LogHelper.getMethodName(callFrame);
                        try {
                            if (callFrame.args[1] instanceof String)
                                addCount(methodName + " - " + callFrame.args[1]);
                            else
                                addCount(methodName);
                        } catch (Throwable throwable) {
                            addCount(methodName);
                        }
                    }
                });
            } catch (Exception e) {

            }
        }
        handler.postDelayed(() -> LogHelper.printMethodCount(map), deadline);
    }

    /**
     * 开始计数
     *
     * @param members  需要计数的方法
     * @param deadline 截止时间
     */
    public static void startCount(Long deadline, Member... members) {
        ArrayList list = new ArrayList();
        Collections.addAll(list, members);
        startCount(deadline, list);
    }

    private static void addCount(String method) {
        String stackClassName = LogHelper.getStackClassName();
        int count = -1;
        try {
            count = map.get(method).get(stackClassName);
        } catch (Exception e) {

        }
        if (count != -1) {
            map.get(method).put(stackClassName, ++count);
        } else {
            HashMap<String, Integer> hasMap = map.get(method);
            if (hasMap == null) hasMap = new HashMap<>();
            hasMap.put(stackClassName, 1);
            map.put(method, hasMap);
        }
    }
}

File: miit_rule_checker/src/main/java/com/loper7/miit_rule_checker/MIITMethodCountChecker.kt
package com.loper7.miit_rule_checker

import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.Log
import top.canyie.pine.Pine
import top.canyie.pine.callback.MethodHook
import java.lang.reflect.Member

/**
 *@Author loper7
 *@Date 2023/11/30 9:27
 *@Description
 **/
object MIITMethodCountChecker {

    private val map : HashMap<String , HashMap<String , Int>> = hashMapOf()
    private val handler by lazy { Handler(Looper.getMainLooper()) }

    /**
     * 开始计数
     * @param members 需要计数的方法
     * @param deadline 截止时间
     */
    @JvmStatic
    fun startCount(members : MutableList<Member?> , deadline : Long = 6000) {
        map.clear()
        for (member in members) {
            if (member == null) continue
            try {
                Pine.hook(member , object : MethodHook() {
                    override fun beforeCall(callFrame : Pine.CallFrame?) {
                        super.beforeCall(callFrame)
                        val methodName = LogHelper.getMethodName(callFrame)
                        try {
                            addCount("$methodName - ${(callFrame !!.args[1] as String)}")
                        } catch (_ : Throwable) {
                            addCount(methodName)
                        }
                    }
                })
            } catch (_ : Exception) {

            }
        }
        handler.postDelayed({
            LogHelper.printMethodCount(map)
        } , deadline)
    }

    /**
     * 开始计数
     * @param members 需要计数的方法
     * @param deadline 截止时间
     */
    fun startCount(vararg members : Member , deadline : Long = 6000) {
        startCount(members.toMutableList() , deadline)
    }

    private fun addCount(method : String) {
        val stackClassName = LogHelper.getStackClassName()
        var count = map[method]?.get(stackClassName) ?: - 1
        if (count != - 1) {
            map[method]?.put(stackClassName , ++ count)
        } else {
            var hasMap = map[method]
            if (hasMap == null) hasMap = hashMapOf()
            hasMap[stackClassName] = 1
            map[method] = hasMap
        }
    }
}